### PR TITLE
Enhance stainnorm luminosity stand

### PR DESCRIPTION
**Pull request by mostafajahanifar**
To be able to perform most of stain normalization algorithms, we need to extract stain matrix from the stained tissue in the image. In other words, the stain matrix should encode only the color information in the tissue region of the image. However, given a target image, that image might include background region as well which can affect the stain matrix and hence the final stain normalization output. Therefore, the first preprocessing step in the stain normalization would be tissue segmentation. 

As we usually use small target image which mostly contain the tissue region, using adaptive thresholding algorithms (like Otsu) would results to imperfect segmentation of target image, for example separating the stromal region from nuclei regions. Therefore, we prefer to use fixed thresholds to separate between background and foreground (tissue) in the image. However, due to the variation in the intensity levels of images, choosing a single threshold may not work for all cases. To overcome this challenge, we propose to first standardize the illumination of the input image and then apply the threshold.

The proposed standardizing method is image contrast enhancement algorithm by linearly adjusting the image intensities. Here, we first find the highest and lowest 2% percentile of the image intensity values, and linearly change all the pixel intensities so that those high and low percentile values become 255 and 0:
`img_out = skimage.exposure.rescale_intensity(img_out, in_range=(p_low, p_high), out_range=(0., 255.))`

This new implementation changes both high and low intensity values in the image and therefore guarantees to achieve a better contrast in the output image in comparison to other method (in stainTool) which adjusts only high percentile of image intensities and would results to a low-contrast image. In the following figure, you can see the effect of each preprocessing method. 

luminosity

As illustrated in the above figure, the luminosity component of the image after stainTool adjustment become very low-contrast that even choosing a high threshold of 0.98 would results to segmenting both background and tissue region (and may miss some tissue parts). On the other hand, in the proposed contrast enhancement method, one can choose a reliable threshold of 0.9 and be sure that only tissue regions are being segmented.

### CODE REVIEW

**Comment by mostafajahanifar** [comment]
> # [Codecov](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=h1) Report
> > Merging [#65](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=desc) ([344dcf5](https://github.com/TIA-Lab/tiatoolbox/commit/344dcf545e92c6253f7331209d986ed8b858a05d)) into [develop](https://codecov.io/gh/TIA-Lab/tiatoolbox/commit/b60bb22cc48f126532d3c015d06f658440cb6828?el=desc) ([b60bb22](https://github.com/TIA-Lab/tiatoolbox/commit/b60bb22cc48f126532d3c015d06f658440cb6828)) will **decrease** coverage by `0.79%`.
> > The diff coverage is `58.33%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=tree)
> 
> ```diff
> @@             Coverage Diff             @@
> ##           develop      #65      +/-   ##
> ===========================================
> - Coverage    88.96%   88.16%   -0.80%     
> ===========================================
>   Files           15       15              
>   Lines          770      769       -1     
>   Branches       111      117       +6     
> ===========================================
> - Hits           685      678       -7     
> - Misses          41       44       +3     
> - Partials        44       47       +3     
> ```
> 
> [Impacted Files](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=tree)	Coverage Δ	
> [tiatoolbox/utils/misc.py](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65/diff?src=pr&el=tree#diff-dGlhdG9vbGJveC91dGlscy9taXNjLnB5)	`73.61% <58.33%> (-3.06%)`	⬇️
> [tiatoolbox/dataloader/wsireader.py](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65/diff?src=pr&el=tree#diff-dGlhdG9vbGJveC9kYXRhbG9hZGVyL3dzaXJlYWRlci5weQ==)	`93.50% <0.00%> (-0.87%)`	⬇️
> [tiatoolbox/dataloader/save_tiles.py](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65/diff?src=pr&el=tree#diff-dGlhdG9vbGJveC9kYXRhbG9hZGVyL3NhdmVfdGlsZXMucHk=)	`93.33% <0.00%> (ø)`	
> [tiatoolbox/cli.py](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65/diff?src=pr&el=tree#diff-dGlhdG9vbGJveC9jbGkucHk=)	`73.18% <0.00%> (+0.19%)`	⬆️
> [tiatoolbox/utils/transforms.py](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65/diff?src=pr&el=tree#diff-dGlhdG9vbGJveC91dGlscy90cmFuc2Zvcm1zLnB5)	`96.29% <0.00%> (+2.17%)`	⬆️
> [Continue to review full report at Codecov](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=continue).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=footer). Last update [b60bb22...2ce9429](https://codecov.io/gh/TIA-Lab/tiatoolbox/pull/65?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

Alright, what should do about this now? @John-P

File: tiatoolbox/utils/misc.py [hunk 1]
```diff
@@ -235,3 +237,35 @@ def mpp2objective_power(mpp):
...
+        >>> img = utils.misc.contrast_enhancer(img, low_p=2, high_p=98)
+
+    """
+    # check if image is not uint8
+    if not img.dtype == np.uint8:
+        raise Exception("Image should be RGB uint8.")
+    img_out = img.copy()
+    p_low, p_high = np.percentile(img_out, (low_p, high_p))
+    if p_low >= p_high:
+        p_low, p_high = np.min(img_out), np.max(img_out)
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:267]: @mostafajahanifar can you check if we can improve code coverage by adding a test?

Comment (mostafajahanifar) [reply tiatoolbox/utils/misc.py:267]: done!

File: fk j [hunk 2]
```diff
@@ -0,0 +1,15 @@
+  [31morigin/HEAD[m -> origin/develop
```
Comment (shaneahmed) [inline fk j:1]: What is the purpose of this file? If it is redundant add to .gitignore

File: tests/test_utils.py [hunk 3]
```diff
@@ -52,3 +52,27 @@ def test_mpp2objective_power(_sample_svs):
 
     with pytest.raises(ValueError):
         utils.misc.mpp2objective_power(mpp=10)
+
+
+def test_contrast_enhancer():
+    """"Test contrast enhancement funcitionality."""
+    input_array = np.array(
```
Comment (shaneahmed) [inline tests/test_utils.py:59]: Can you elaborate in comments how you chose the values to run the tests?

File: tiatoolbox/utils/misc.py [hunk 4]
```diff
@@ -235,3 +237,35 @@ def mpp2objective_power(mpp):
         # Double the objective power as mpp halves
         return 10 * 2 ** (np.ceil(np.log2(0.15 / mpp)) + 2)
     raise ValueError()
+
+
+def contrast_enhancer(img, low_p=2, high_p=98):
+    """Enhancing contrast of the input RGB image using intensity adjustment.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:243]: Add the method to [usage.rst](https://github.com/TIA-Lab/tiatoolbox/blob/develop/docs/usage.rst) so it is compiled on readthedocs.

**Review by shaneahmed — COMMENTED** [review]
Thanks @mostafajahanifar. Looks great. Please address the comments I added and I will approve the PR. Please check with @simongraham if he has any comments.

File: tests/test_utils.py (hunk 3 above)
Comment (mostafajahanifar) [reply tests/test_utils.py:59]: The goal is to check if `contrast_enhancer` method can reproduce results for uint8 images with 3 channels. Qualitative evaluation of the method has been done several times using example images. To check the performance and reproducibility in a quantitative way, and also to reduce uploading more images to the repository, a random tensor is generated and fed into the method. The `output_array ` should be exactly same as expected `result_array` (which has been numerically validated offline).

File: fk j (hunk 2 above)
Comment (mostafajahanifar) [reply fk j:1]: I'm not sure, it seems to be a incomplete redundant file generated by git. I will remove it completely and add it to .gitignore as well.
Thanks for noting.

File: tests/test_utils.py (hunk 3 above)
Comment (shaneahmed) [reply tests/test_utils.py:59]: Please can you add a short description as comments in the test itself? So anyone modifying or writing the code may understand it easily.

Comment (mostafajahanifar) [reply tests/test_utils.py:59]: yes sure.

File: fk j (hunk 2 above)
Comment (shaneahmed) [reply fk j]: Removed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/b7e39b24199f9e1270b0e27da304e71427c31582

File: tests/test_utils.py (hunk 3 above)
Comment (shaneahmed) [reply tests/test_utils.py:63]: Addressed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/cf87a50d92e62ea2597f9c4da77f9dcf2cc2491e

File: tiatoolbox/utils/misc.py (hunk 4 above)
Comment (shaneahmed) [reply tiatoolbox/utils/misc.py:247]: Addressed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/d9752c6449023ca08f2e7a84bada8bc4130179d1

**Review by shaneahmed — APPROVED** [review]

**Pull request merged**